### PR TITLE
DI Modules 7063

### PR DESCRIPTION
**Pull request by dwas01**
DI Modules 7063

### CODE REVIEW

File: Modules/ApplicationInsights/2.0/deploy.json [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    },
+    "appInsightsAppId": {
+      "type": "string",
+      "value": "[reference(variables('appInsightsresourceId'),variables('apiVersion')).AppId]",
+      "metadata": {
+        "description": "Application Insights Paalication Id"
+      }
+    },
+    "appInsightsStorageAccountName": {
```
Comment (jorgecotillo) [inline Modules/ApplicationInsights/2.0/deploy.json]: I think this is not needed, you are combining two resources in the same template, the idea is to create one resource (and its nested or related resources -> e.g. VirtualNetwork and Subnets) in the same template another example is VirtualMachine and NetworkInterface, NetworkInterface is always used by a VirtualMachine, so in this case we place both resources in the same template. In this specific case a Storage Account can be used by different resources (diagnostics, artifacts, store bootstrap blob containers for Palo Alto, etc.)

File: Modules/ApplicationInsights/2.0/deploy.json [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  "variables": {
+    "appInsightsName": "[parameters('appInsightsName')]",
+    "appInsightsresourceId": "[resourceId('Microsoft.Insights/components',parameters('appInsightsName'))]",
+    "apiVersion": "[providers('Microsoft.Insights','components').apiVersions[0]]",
+    "storageAccountName": "[parameters('storageAccountName')]",
+    "pidName": "[concat('pid-',parameters('cuaId'))]"
+  },
+  "resources": [
+    {
+      "condition": "[not(empty(parameters('cuaId')))]",
```
Comment (jorgecotillo) [inline Modules/ApplicationInsights/2.0/deploy.json]: @bennage do we want to include this value in every single template? I think is a good idea if and only if we add a condition to prevent the deployment of this resource. This resource is related to the Partner Tracking ACR ID.

File: Modules/CognitiveServices/2.0/deploy.json [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        }
+      }
+    },
+    {
+      "type": "Microsoft.CognitiveServices/accounts",
+      "apiVersion": "[providers('Microsoft.CognitiveServices','accounts').apiVersions[0]]",
+      "name": "[parameters('accountName')]",
+      "kind": "TextTranslation",
+      "location": "[parameters('location')]",
+      "tags": {
```
Comment (jorgecotillo) [inline Modules/CognitiveServices/2.0/deploy.json]: Same as above, depending on what we agree, we might need to make the "tags" value to be a condition, as mentioned above.

File: Modules/EventHub/2.0/deploy.json [hunk 4]
```diff
@@ -1,140 +1,324 @@
 {
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: @senthilkungumaraj did you guys come to an agreement with this module?

File: Modules/EventHub/2.0/deploy.json [hunk 5]
```diff
@@ -1,140 +1,324 @@
...
+          "resources": []
+        }
+      }
+    },
+    {
       "type": "Microsoft.EventHub/namespaces",
+      "apiVersion": "[variables('apiVersion')]",
+      "name": "[parameters('namespaceName')]",
       "location": "[parameters('location')]",
+      "tags": {
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: Same as above, we might need to move this value as an input parameter

File: Modules/EventHub/2.0/deploy.json [hunk 6]
```diff
@@ -1,140 +1,324 @@
...
+        "Application": "[parameters('tagApplication')]",
+        "Project": "[parameters('tagProject')]",
+        "Owner": "[parameters('tagOwner')]",
+        "OwnerEmail": "[parameters('tagOwnerEmail')]"
+      },
       "sku": {
-        "name": "[parameters('eventHubSku')]",
-        "tier": "[parameters('eventHubSku')]",
-        "capacity": "[parameters('skuCapacity')]"
+        "name": "Standard",
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: Make sure to use parameters for name, tier and capacity instead of fixed values to provide flexibility to module consumers.

File: Modules/EventHub/2.0/deploy.json [hunk 7]
```diff
@@ -1,140 +1,324 @@
...
+      "dependsOn": [
+        "[resourceId('Microsoft.EventHub/namespaces/eventhubs', parameters('namespaceName'), parameters('eventHubName'))]",
+        "[variables('namespaceResourceId')]"
+      ],
+      "properties": {}
     }
   ],
   "outputs": {
-    "diagnosticEventHubName": {
+    "namespaceName": {
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: I am in favor of this generic name, because EventsHub can be used for different purposes and not only diagnostics.

File: Modules/MachineLearning/2.0/deploy.json [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        "mode": "Incremental",
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "resources": []
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Storage/storageAccounts",
```
Comment (jorgecotillo) [inline Modules/MachineLearning/2.0/deploy.json]: Same as above, this resource is not related to Machine Learning, instead Machine Learning uses a Storage Account, since we already have a Storage Accounts module, let's pass the Storage Account Id instead.

File: Modules/EventHub/2.0/deploy.json [hunk 9]
```diff
@@ -1,140 +1,324 @@
...
+      "defaultValue": "",
+      "metadata": {
+        "description": "Optional. The Email address of the business owner for the application"
       }
     }
   },
   "variables": {
+    "namespaceName": "[parameters('namespaceName')]",
+    "namespaceResourceId": "[resourceId('Microsoft.EventHub/Namespaces',parameters('namespaceName'))]",
+    "diagnosticStorageAccountId": "[parameters('diagnosticStorageAccountId')]",
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: one to one mapping? make sure to use the parameter instead.

File: Modules/EventHub/2.0/deploy.json [hunk 10]
```diff
@@ -1,140 +1,324 @@
...
+    "diagnosticWorkspaceId": "[parameters('logAnalyticsWorkspaceId')]",
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: same as above.

File: Modules/EventHub/2.0/deploy.json [hunk 11]
```diff
@@ -1,140 +1,324 @@
...
-        "tier": "[parameters('eventHubSku')]",
-        "capacity": "[parameters('skuCapacity')]"
+        "name": "Standard",
+        "tier": "Standard",
+        "capacity": 1
       },
-      "tags": {
-        "displayName": "[parameters('namespaceName')]"
+      "properties": {
+        "zoneRedundant": false,
```
Comment (jorgecotillo) [inline Modules/EventHub/2.0/deploy.json]: I feel like these values must come as input parameters and we can set defaults if we'd like, try to limit the use of fixed values.

File: Modules/MachineLearning/2.0/deploy.json [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+      "dependsOn": [
+        "[variables('storageAccountResourceId')]"
+      ]
+    },
+    {
+      "type": "Microsoft.MachineLearning/commitmentPlans",
+      "apiVersion": "[providers('Microsoft.MachineLearning', 'commitmentPlans').apiVersions[0]]",
+      "name": "[parameters('planName')]",
+      "location": "[parameters('location')]",
+      "tags": {
```
Comment (jorgecotillo) [inline Modules/MachineLearning/2.0/deploy.json]: Same as above, we might need to pass tags object as an input parameter

File: Modules/ApplicationInsights/2.0/deploy.json [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "resources": []
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Insights/components",
+      "apiVersion": "[providers('Microsoft.Insights','components').apiVersions[0]]",
```
[the remark this answers is not in the document]
Comment (bennage) [reply Modules/ApplicationInsights/2.0/deploy.json]: As a rule, I won't to avoid ambiguous versions (i.e., latest) in the templates. 
We should use an explicit api versions and rely on tagging in the repo to handle backcompat issues.

File: Modules/ApplicationInsights/2.0/deploy.json [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+        "ApplicationId": "[parameters('appInsightsName')]",
+        "Application_Type": "[parameters('appInsightsType')]"
+      }
+    },
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "apiVersion": "[providers('Microsoft.Storage','storageAccounts').apiVersions[0]]",
+      "name": "[parameters('storageAccountName')]",
+      "location": "[resourceGroup().location]",
+      "tags": {
```
[the remark this answers is not in the document]
Comment (bennage) [reply Modules/ApplicationInsights/2.0/deploy.json]: Agreed.

File: Modules/ApplicationInsights/2.0/deploy.json [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        "Project": "[parameters('tagProject')]",
+        "Owner": "[parameters('tagOwner')]",
+        "OwnerEmail": "[parameters('tagOwnerEmail')]"
+      },    
+      "properties": {
+        "ApplicationId": "[parameters('appInsightsName')]",
+        "Application_Type": "[parameters('appInsightsType')]"
+      }
+    },
+    {
```
[the remark this answers is not in the document]
Comment (dwas01) [reply Modules/ApplicationInsights/2.0/deploy.json]: Storage Account Removed

File: Modules/ApplicationInsights/2.0/deploy.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (dwas01) [reply Modules/ApplicationInsights/2.0/deploy.json]: Each module now has a variable for APIVersion ...

File: Modules/ApplicationInsights/2.0/deploy.json (hunk 14 above)
[the remark this answers is not in the document]
Comment (dwas01) [reply Modules/ApplicationInsights/2.0/deploy.json]: Change the code to 

 "tagValues": {
      "type": "object",
      "defaultValue": "",
      "metadata": {
        "description": "Optional. Azure Resource Tags object"
      }
    }


"tags": "[if(empty(parameters('tagValues')), json('null'), parameters('tagValues'))]",
  Tags are now optional

File: Modules/ApplicationInsights/2.0/deploy.json (hunk 1 above)
Comment (dwas01) [reply Modules/ApplicationInsights/2.0/deploy.json]: removed

File: Modules/ApplicationInsights/2.0/deploy.json (hunk 2 above)
Comment (dwas01) [reply Modules/ApplicationInsights/2.0/deploy.json]: In every module

File: Modules/CognitiveServices/2.0/deploy.json (hunk 3 above)
Comment (dwas01) [reply Modules/CognitiveServices/2.0/deploy.json]: tags changed to new format

File: Modules/EventHub/2.0/deploy.json (hunk 4 above)
Comment (dwas01) [reply Modules/EventHub/2.0/deploy.json]: Done we sync'd

File: Modules/EventHub/2.0/deploy.json (hunk 6 above)
Comment (dwas01) [reply Modules/EventHub/2.0/deploy.json]: Added parameters

File: Modules/EventHub/2.0/deploy.json (hunk 5 above)
Comment (dwas01) [reply Modules/EventHub/2.0/deploy.json]: tags changed to new format

File: Modules/EventHub/2.0/deploy.json (hunk 7 above)
Comment (dwas01) [reply Modules/EventHub/2.0/deploy.json]: Name changed

File: Modules/MachineLearning/2.0/deploy.json (hunk 8 above)
Comment (dwas01) [reply Modules/MachineLearning/2.0/deploy.json]: Removed and StorageResourceId Parameter now required

File: Modules/MachineLearning/2.0/deploy.json (hunk 12 above)
Comment (dwas01) [reply Modules/MachineLearning/2.0/deploy.json]: tags now new format
